### ISSUE TITLE
Request: Sort out functions etc.

### ISSUE THREAD

**Issue by r3incarnat0r** [Disassembler0/Win10-Initial-Setup-Script#115]
In case a bigger update such as #81 and #99 might be planned, I'd also like to suggest some more changes on this occasion. I can already see you won't like it, as it would change too much. :smiley:  For now I just put a list of ideas here (including duplicated ideas from #81 and #99 to have a summary) and leave reasons and arguments for later, if you'd be even willing to consider such changes (or asking why :wink: ).

- as requsted in #99, move tweaks from main script and even more, regroup them to separate modules based on their scope
- modules specified by arguments or load .\Modules\*.psm1 if unspecified
- move default preset from main script to a text file keeping this well arranged form with comments and also sort them out to have scope groups separated (while keeping them sorted thematically as they already are now)
- rename *.preset to *.txt or *.ini
- presets specified by arguments or also maybe use (scriptname).(presetextension) if unspecified and such file is found (eg. Win10.txt, if running Win10.ps1 directly).
- rename tweak functions respecting powershell naming convention (eg. Set-Something, Disable-Feature,...)
- while we have GPO tweaks in their own module, have their non-GPO versions in other modules, differ GPO functions by modified name (eg. Disable-Something vs. GPO-Disable-Something) (looks weird, but we can work in it), update echos in GPO versions (eg. Disabling Something (GPO)...)
- self-descriptive directory scheme example:
```
 .
 |-[Modules]
 |  |-LocalTweaks.psm1
 |  |-UserTweaks.psm1
 |  |-GPO.psm1
 |-Default.cmd
 |-Default.txt
 |-Win10.ps1
 |
 etc.
```
The whole circus above is mostly meant to simplify future code additions and to have ability adding more and more tweaks without making a mess. Thanks!

**Comment by Disassembler0** [comment]
You're right. I don't like it. :)

> move tweaks from main script and even more, regroup them to separate modules based on their scope

I can understand the separation of tweaks from the main script into a kind of library, but I don't see any practical reason why to split them even more. The project started as a single stupid file which anyone could store on their USB stick with Windows installer and be done with it. No extra files, no dependencies, however not much modularity either. So later on I have added the option for presets which would add some modularity, but made it in such a way so it's still possible just to have a single file.

Now you and bunch of others persuaded me into splitting the script into two - one with the actual tweaks and one with the main program loop. The original script doesn't benefit from this in any way, because it will still work the same. It's actually a disadvantage for the original script because now you will have to have carry multiple files, but *some other* scripts which I absolutely don't have to care about could benefit from that. I can't say I'm 100% OK with that, but I understand. However I don't understand who and why could benefit from splitting them even more.

> modules specified by arguments or load .\Modules*.psm1 if unspecified

I despise any implicitness, introspection or any other means which try to be smarter than the user and often turn out to be a problem. Different people tend tend to expect different things, because it makes sense to *them*. So that's a definite no. If you want anything to be loaded, you have to say so explicitly.

> move default preset from main script to a text file...

That's already partially done. The default preset exists as a separate file and part of ideas in #99 is to strip the main loop to the bare minimum, sou you'll need to have at least 3 files for anything to work (tweak library, preset and the main loop). So yeah, that will be done eventually.

> ...with comments...

There already are comments in the tweak library. The purpose of the preset should be just to pick which of them should be run. I don't want to maintain the same comments on multiple places and if I have to decide whether to keep them in the code or move them from the code to the default preset, I'd rather keep them in the code.

> ...sort them out to have scope groups separated...

I don't understand what do you mean by this part. I try to keep them sorted even within the group, but it's not always obvious which tweaks relate to each other.

> rename *.preset to *.txt or *.ini

The name is completely arbitrary. You can even rename it to **.exe* and as long as it's a plaintext file, the script will read it. For sure it's not **.ini* as it doesn't have any sections (and adding them just for sake of correctness of the format would server no purpose). And I originally didn't want to keep **.txt* as people tend to be blind towards text files and files without extensions.

> presets specified by arguments or also maybe use (scriptname).(presetextension) if unspecified and such file is found (eg. Win10.txt, if running Win10.ps1 directly).

Nothing more to say here. See the "*move default preset...*" and "*modules specified by arguments...*" parts.

> rename tweak functions respecting powershell naming convention

Although you might be right about this one (it will be already a full-featured PowerShell module, it should stick to the conventions), the problem is that the functions themselves use completely different paradigm. The PowerShell way of writing `DisableTelemetry` would be something like `Set-Telemetry -Status Disabled`, which, although not entirely impossible, would be very difficult to implement in the similar fashion as it is now. So normally, when you want to call such `Set-Telemetry -Status Disabled` as a single function without parameters, you make a wrapper for it. And you can call the wrapper `DisableTelemetry` :P Besides I'm pretty sure people would really hate me for this one because it would invalidate all existing presets in their entirety.

> while we have GPO tweaks in their own module, have their non-GPO versions in other modules...

![No](https://media.giphy.com/media/2Tn10AXN6B98k/giphy.gif)

You do realize I don't just write the code, but I even test it, right? :) It would be really huge pain in the buttocks to find and hunt down all possible conflicts. Moreover, my dream would be to have *everything* implemented in GPO but some settings don't have any appropriate policy. On the other hand, there is also a bunch of tweaks which can be applied *only* through GPO because the non-GPO way has the values stored in some data blobs unique for the given computer / user or in some proprietary data store completely outside of registry.

> The whole circus above is mostly meant to simplify future code additions and to have ability adding more and more tweaks without making a mess.

Well, that's already implied by #81 and #99. Like I said - even though I'm stubborn and don't like it, eventually you will end up with one (or more, if you have custom) PS module, one or more presets, the `Win10.ps1` script and one batch file to tie them all together. So it could look like this:
```
 .
 |-MyBasic.preset
 |-MyTweaks.psm1
 |-MyWorkstation.cmd
 |-MyWorkstation.preset
 |-Win10.ps1
 |-Win10.psm1
```
Where the  `MyWorkstation.cmd` would contain
```
powershell.exe -NoProfile -ExecutionPolicy Bypass ^
    -File "%~dp0Win10.ps1" ^
    -Include "%~dp0Win10.psm1" ^
    -Include "%~dp0MyTweaks.psm1" ^
    -Preset "%~dp0MyBasic.preset" ^
    -Preset "%~dp0MyWorkstation.preset"
```

From these, all the files with names starting with "*My**" are completely in your power. And if you want to contribute, you'll most likely want to change *Win10.psm1*. Then it doesn't really matter if it's one big file or a bunch of small ones. As long as it's not a file per tweak (No! Don't even think about it! :) ) you'll always make the same amount of "mess".

**Comment by Disassembler0** [comment]
Implemented in 1e977ab95fe04a99cdd152d342586de01d0aec96. Probably.
